File: 30days/12/12.php
<?php

class person {
    protected $firstName, $lastName, $id;
    public function __construct($first_name, $last_name, $identification) {
        $this->firstName = $first_name;
        $this->lastName = $last_name;
        $this->id = $identification;
    }
    function printPerson() {
        print("Name: {$this->lastName}, {$this->firstName}\n");
        print("ID: {$this->id}\n");
    }
}

class Student extends person {
    private $testScores;
    public function __construct($first_name, $last_name, $identification, $scores) {
        parent::__construct($first_name, $last_name, $identification);
        $this->testScores = $scores;
    }

    public function calculate() {
        $grade_tot = 0;
        for ($i = 0; $i < count($this->testScores); $i++) {
            $grade_tot += $this->testScores[$i];
        }

        $result = $grade_tot / count($this->testScores);

        if ($result >= 90) {
            return "O";
        } else if ($result >= 80) {
            return "E";
        } else if ($result >= 70) {
            return "A";
        } else if ($result >= 55) {
            return "P";
        } else if ($result >= 40) {
            return "D";
        } else {
            return "T";
        }
    }
}

$file = fopen("php://stdin", "r");

$name_id = explode(' ', fgets($file));

$first_name = $name_id[0];
$last_name = $name_id[1];
$id = (int)$name_id[2];

fgets($file);

$scores = array_map(intval, explode(' ', fgets($file)));

$student = new Student($first_name, $last_name, $id, $scores);

$student->printPerson();

print("Grade: {$student->calculate()}");

File: 30days/12/12.rb
#!/usr/bin/env ruby

class Person
    def initialize(firstName, lastName, id)
        @firstName = firstName
        @lastName = lastName
        @id = id
    end

    def printPerson()
        print("Name: ",@lastName , ", " + @firstName ,"\nID: " , @id)
    end
end

class Student < Person
    def initialize(first_name, last_name, id, scores)
        super(first_name, last_name, id)
        @scores = scores
    end

    def calculate
        result = 0
        @scores.each do |score|
            result += score
        end

        result /= @scores.length
        return "O" if result >= 90
        return "E" if result >= 80
        return "A" if result >= 70
        return "P" if result >= 55
        return "D" if result >= 40
        return "T"
    end
end

input = gets.split()
firstName = input[0]
lastName = input[1]
id = input[2].to_i
numScores = gets.to_i
scores = gets.strip().split().map!(&:to_i)
s = Student.new(firstName, lastName, id, scores)
s.printPerson
print("\nGrade: " + s.calculate)
